### ISSUE TITLE
Sequence pattern sometimes matches strings

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#11099]
Pyright version: 1.1.407, Python version: 3.14

Pyright understands that `str` and `bytes` objects don't match sequence patterns in the `match` statement:

```py
def foo(x: str | bool) -> int:
    match x:
        case [*xs]: return 100  # reportUnnecessaryComparison: Pattern will never be matched for subject type "str | bool"
        case _: return 200
```

However, that does not happen when `str` (or `bytes`) is part of a larger sequence type:
```py
from collections.abc import Sequence

def foo(x: Sequence[object] | bool) -> int:
    match x:
        case [*_]: return 100
        case True: return 200
        case False: return 300
        # if `x` is a `str` or `bytes` object, no patterns will be matched here

y = foo("banana")
```
`pyright` does not raise any issues with the `foo` function. `y` is inferred to be `int`, but is `None` at runtime.
